src/core/thread/csl_tx_scheduler.hpp

```diff
@@ -0,0 +1,215 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef CSL_TX_SCHEDULER_HPP_
+#define CSL_TX_SCHEDULER_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/locator.hpp"
+#include "common/message.hpp"
+#include "common/time.hpp"
+#include "mac/mac.hpp"
+#include "mac/mac_frame.hpp"
+#include "thread/indirect_sender_frame_context.hpp"
+
+namespace ot {
+
+/**
+ * @addtogroup core-mesh-forwarding
+ *
+ * @brief
+ *   This module includes definitions for CSL transmission scheduling.
+ *
+ * @{
+ */
+
+#if OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+
+class Child;
+
+/**
+ * This class implements CSL tx scheduling functionality.
+ *
+ */
+class CslTxScheduler : public InstanceLocator
+{
+    friend class Mac::Mac;
+    friend class IndirectSender;
+
+public:
+    enum
+    {
+        kMaxCslTriggeredTxAttempts     = OPENTHREAD_CONFIG_MAC_MAX_TX_ATTEMPTS_INDIRECT_POLLS,
+        kCslFrameRequestAheadThreshold = 2000 / kUsPerTenSymbols,
+    };
+
+    /**
+     * This class defines all the child info required for scheduling CSL transmissions.
+     *
+     * `Child` class publicly inherits from this class.
+     *
+     */
+    class ChildInfo
+    {
+    public:
+        uint8_t GetCslTxAttempts(void) const { return mCslTxAttempts; }
+        void    SetCslTxAttempts(uint8_t aCslTxAttempts) { mCslTxAttempts = aCslTxAttempts; }
+        void    IncrementCslTxAttempts(void) { mCslTxAttempts++; }
+        void    ResetCslTxAttempts(void) { SetCslTxAttempts(0); }
+
+        bool IsCslSynchronized(void) const { return mCslSynchronized && mCslPeriod > 0; }
+        void SetCslSynchronized(bool aCslSynchronized) { mCslSynchronized = aCslSynchronized; }
+
+        uint8_t GetCslChannel(void) const { return mCslChannel; }
+        void    SetCslChannel(uint8_t aChannel) { mCslChannel = aChannel; }
+
+        uint32_t GetCslSyncTimeout(void) const { return mCslSyncTimeout; }
+        void     SetCslSyncTimeout(uint32_t aTimeout) { mCslSyncTimeout = aTimeout; }
+
+        uint16_t GetCslPeriod(void) const { return mCslPeriod; }
+        void     SetCslPeriod(uint16_t aPeriod) { mCslPeriod = aPeriod; }
+
+        uint16_t GetCslPhase(void) const { return mCslPhase; }
+        void     SetCslPhase(uint16_t aPhase) { mCslPhase = aPhase; }
+
+        TimeMilli GetCslLastHeard(void) const { return mCslLastHeard; }
+        void      SetCslLastHeard(TimeMilli aCslLastHeard) { mCslLastHeard = aCslLastHeard; }
+
+    private:
+        uint8_t   mCslTxAttempts : 7;   // Number of csl triggered tx attempts.
```

Comment:
nit: for consistency
```suggestion
        uint8_t   mCslTxAttempts : 7;   ///< Number of csl triggered tx attempts.
```